lib/connectors/mongodb/connector.rb

```diff
@@ -56,30 +60,114 @@ def initialize(configuration: {})
       end
 
-      def yield_documents
+      def yield_documents(&on_doc_serialization)
+        check_filter_config
+
+        call_db_function_on_collection(&on_doc_serialization)
+      end
+
+      private
+
+      def create_db_cursor_on_collection(collection)
+        return create_find_cursor(collection) if @active_filter_config[:find].present?
+
+        return create_aggregate_cursor(collection) if @active_filter_config[:aggregate].present?
+
+        collection.find
+      end
+
+      def check_filter_config
+        return unless filtering_present?
+
+        check_find_and_aggregate
+      end
+
+      def check_find_and_aggregate
+        invalid_keys_msg = "Only one of #{ALLOWED_TOP_LEVEL_FILTER_KEYS} is allowed in the filtering object. Keys present: '#{@active_filter_config.keys}'."
+        raise Utility::InvalidFilterConfigError.new(invalid_keys_msg) if @active_filter_config.keys.size != 1
+      end
+
+      def call_db_function_on_collection(&on_doc_serialization)
         with_client do |client|
           # We do paging using skip().limit() here to make Ruby recycle the memory for each page pulled from the server after it's not needed any more.
           # This gives us more control on the usage of the memory (we can adjust PAGE_SIZE constant for that to decrease max memory consumption).
           # It's done due to the fact that usage of .find.each leads to memory leaks or overuse of memory - the whole result set seems to stay in memory
           # during the sync. Sometimes (not 100% sure) it even leads to a real leak, when the memory for these objects is never recycled.
-          cursor = client[@collection].find
+          cursor, options = create_db_cursor_on_collection(client[@collection])
           skip = 0
 
+          found_overall = 0
+
+          # if no overall limit is specified by filtering use -1 to not break ingestion, when no overall limit is specified (found_overall is only increased,
+          # thus can never reach -1)
+          overall_limit = -1
+
+          if options_present?(options)
+            # there could be a skip parameter defined for filtering
+            skip = Utility::Common.return_if_present(options[:skip], 0)
+            # there could be a limit parameter defined for filtering -> used for an overall limit (not a page limit, which was introduced for memory optimization)
+            overall_limit = Utility::Common.return_if_present(options[:limit], -1)
+          end
+
+          overall_limit_reached = false
+
           loop do
-            found_count = 0
+            found_in_page = 0
+
             view = cursor.skip(skip).limit(PAGE_SIZE)
             view.each do |document|
-              yield serialize(document)
-              found_count += 1
+              yield serialize(document) { on_doc_serialization.call if block_given? }
+
+              found_in_page += 1
+              found_overall += 1
+
+              overall_limit_reached = found_overall == overall_limit
```

Comment:
Good point, but I think this will only work, if we also check, that the `overall_limit` is not set -1 (flagging, that there's no overall limit set). Probably `Float::INFINITY` makes more sense semantically as a default value for the overall limit.